content/posts/observability-in-the-cloud.md

```diff
@@ -0,0 +1,86 @@
+Finally I roughly outline how it should be approached.
+
+## Observability vs. Monitoring
+
+For the difference between monitoring and observability, I recommend Cindy Sridharan's book Distributed Systems Observability and Charity Major's various blog posts on the topic (e.g. [Observability is a Many-Splendored Definition](https://charity.wtf/2020/03/03/observability-is-a-many-splendored-thing/)) who go into more detail.
+
+Here, I summarize the main points and draw some conclusions.
+
+Monitoring is what most of us are used to.
+It encompasses collection of telemetry (metrics, logs and traces), visualization through dashboards and alerting, with metrics being the primary data type used.
+Monitoring is primarily the process of collecting telemetry and acting upon it.
+It deals with the known-unknowns (e.g. we know that we may run out of memory or disk and can monitor it, but we do not precisely know when and because of which processes it happens).
+
+Observability, on the other side, describes a property of a system.
+A system is observable, if you can deduce its internal state (such as the value of its internal variables and position in the callstack), from the outputs you can observe.
+These outputs can be specific behaviours exhibited to users, and in practice rely heavily on telemetry we collect from the system's environment (such as the machine it's running on) or deliberately output from the system.
+A system can be described as "more observable", when we can more accurately describe its internal state, and fully observable, when we can always describe its internal state with 100% accuracy.
+Making a system observable usually requires using monitoring tools and processes, but it allows us to deal with a larger set of problems, namely, the unkown-unknowns, and effectively troubleshoot issues that we could not really imagine before they arose.
+
+## The Limits of Monitoring
+
+So, now that we know the difference between monitoring and observability, how did monitoring help us deal with problems in the past?
+
+Once something goes wrong, we may use the alerts and dashboards to get an overview and over which application, machine, or component is affected.
+Additionally, we head into the logs, to try and figure out where exactly things went wrong.
+If it is a known issue, we can usually apply a known fix.
+But if it's an unknown issue, we often have to dig through extensive logs, manually correlate them and finally we may start up the application with a debugger, and try to reproduce the exact conditions.
+We iteratively go back to the logs, to grab more pieces of evidence and approximate the state of the system when the error occurred.
+Telemetry in our monitoring was usually highly aggregated already, but this is a minor issue, since most problems are clearly attributable to a specific system and we can effectively use a debugger to investigate it.
```

Comment:
The point of this sentence is not entirely clear to me. Could this be rephrased somehow?

Reply:
I rephrased it slightly as well as the section after it.